### PR TITLE
LinearAlgebra: Handle more types as zero preserving

### LINKED ISSUE #46292: `map` for `Diagonal` matrix is inefficient
Applying a function to a `Diagonal` matrix can convert it to a dense matrix without need:

```Julia
julia> using LinearAlgebra
julia> using StaticArrays
julia> map(SMatrix{1,1}, Diagonal([1, 2]))
2×2 Matrix{SMatrix{1, 1, Int64, 1}}:
 [1;;]  [0;;]
 [0;;]  [2;;]
```
The result of `map` is a dense matrix (`Matrix`) instead of a diagonal matrix (`Diagonal`).

Although `iszero(SMatrix{1,1}(1))`, `map` converts the matrix to a dense matrix.

### PR DESCRIPTION
**Pull request by eschnett**
Closes https://github.com/JuliaLang/julia/issues/46292

### CODE REVIEW

File: stdlib/LinearAlgebra/src/structuredbroadcast.jl [hunk 1]
```diff
@@ -126,6 +126,7 @@ fails as `zero(::Tuple{Int})` is not defined. However,
 """
 iszerodefined(::Type) = false
 iszerodefined(::Type{<:Number}) = true
+iszerodefined(::Type{<:AbstractArray}) = true
```
Comment (vchuravy) [inline stdlib/LinearAlgebra/src/structuredbroadcast.jl:129]: Isn't this dependent on the `eltype` of the AbstractArray?

Comment (eschnett) [reply stdlib/LinearAlgebra/src/structuredbroadcast.jl]: Yes. Done.

**Comment by eschnett** [comment]
The CI errors are in `InteractiveUtils` and `threads`, and are unrelated to this PR. This is ready for review.

**Comment by eschnett** [comment]
@vchuravy Could you review this again?

**Comment by eschnett** [comment]
@vchuravy Could you review this again? The CI errors appear unrelated.

**Review by dkarrasch — APPROVED** [review]

**Pull request merged**